src/main/java/zack/util/Ui.java

```diff
@@ -0,0 +1,205 @@
+     * Prints the goodbye message when the user exits the application.
+     */
+    public String showGoodbyeMessage() {
+        return "Goodbye! Hope to see you again soon!";
+    }
+
+    /**
+     * Shows a message when a task is marked as done or undone.
+     *
+     * @param task   The task that was marked.
+     * @param isDone True if the task is marked as done, false otherwise.
+     */
+    public String showMarkedTask(Task task, boolean isDone) {
+        if (isDone) {
+            return "Nice! I've marked this task as done:\n" + task;
+        } else {
+            return "OK, I've marked this task as not done yet:\n" + task;
+        }
+    }
+
+    /**
+     * Displays a list of tasks to the user.
+     *
+     * @param tasks The list of tasks to display.
+     */
+    public String showTaskList(ArrayList<Task> tasks) {
+        if (tasks.isEmpty()) {
+            return "The task list is currently empty. Try adding some tasks!";
+        }
+        StringBuilder sb = new StringBuilder();
+        sb.append("Here are the tasks in your list:\n");
+        for (int i = 0; i < tasks.size(); i++) {
+            sb.append((i + 1)).append(".").append(tasks.get(i)).append("\n");
+        }
+        return sb.toString();
+    }
+
+    /**
+     * Displays tasks on a specific date to the user.
+     *
+     * @param date  The date for which tasks are displayed.
+     * @param tasks The list of tasks happening on the specified date.
+     */
+    public String showTasksOnDate(LocalDate date, ArrayList<Task> tasks) {
+        if (tasks.isEmpty()) {
+            return "It seems that are no tasks that correspond to this date. Try another date?";
+        }
+        StringBuilder sb = new StringBuilder();
+        sb.append("Tasks on ").append(date.format(DateTimeFormatter.ofPattern("MMM dd yyyy"))).append(":\n");
+        for (Task task : tasks) {
+            sb.append(task).append("\n");
+        }
+        return sb.toString();
+    }
+
+    /**
+     * Displays an error message when loading tasks fails.
+     *
+     * @param e The exception that occurred while loading tasks.
+     */
+    public void showLoadingError(ZackException e) {
+        System.err.println("Error: " + e.getMessage()); // Display the exception message
+    }
+
+    /**
+     * Displays a list of tasks that have been found based on a search operation.
+     *
+     * This method is used to display a list of tasks that match a search criteria
+     * in the specified TaskList. If no matching tasks are found, it prints a message
+     * indicating that no tasks were found. If matching tasks are found, it prints a
+     * numbered list of those tasks with their details.
+     *
+     * @param foundTasks The TaskList containing the matching tasks to display.
+     * @throws ZackException If there is an error or exception while displaying
+     *                      the found tasks.
+     */
+    public String showFoundTasks(TaskList foundTasks) throws ZackException {
```

Comment:
I like how the function names are detailed and self-explanatory. Makes the code easy to understand.